### PR TITLE
 Use a new file to exchange status with isotovideo (isotovideo process might already be finished at query time)

### PR DESCRIPTION
**Pull request by perlpunk**
Results for tests running less than 10 seconds aren't displayed in the webui, and the job is marked as failed.
This happens because the status call to isotovideo is happening too late, when isotovideo is already finished, so it doesn't get anything back.

Instead we are reading the status from a file that will survive the isotovideo process.
See https://github.com/os-autoinst/os-autoinst/pull/1212

* Remove 'backend': isotovideo status doesn't return this

Depends on https://github.com/os-autoinst/os-autoinst/pull/1212

See also:
* https://progress.opensuse.org/issues/39845

### CODE REVIEW

File: lib/OpenQA/Worker/Job.pm [hunk 1]
```diff
@@ -660,35 +662,40 @@ sub _upload_results {
...
         worker_id             => $worker_id,
         cmd_srv_url           => $job_url,
         worker_hostname       => $global_settings->{WORKER_HOSTNAME},
-        test_execution_paused => $status_from_os_autoinst->{test_execution_paused} // 0,
+        test_execution_paused => 0,
     );
 
+    my $test_status = {};
+    if (open my $fh, '<', $status_file) {
+        $test_status = Cpanel::JSON::XS->new->pretty(1)->decode(
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply lib/OpenQA/Worker/Job.pm]: good catch - no I guess it doesn't ;-)

File: lib/OpenQA/Worker/Job.pm [hunk 2]
```diff
@@ -660,35 +662,40 @@ sub _upload_results {
...
     my %status          = (
         worker_id             => $worker_id,
         cmd_srv_url           => $job_url,
         worker_hostname       => $global_settings->{WORKER_HOSTNAME},
-        test_execution_paused => $status_from_os_autoinst->{test_execution_paused} // 0,
+        test_execution_paused => 0,
     );
 
+    my $test_status = {};
+    if (open my $fh, '<', $status_file) {
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply lib/OpenQA/Worker/Job.pm]: ok

[the remark this answers is not in the document]
Comment (perlpunk) [reply lib/OpenQA/Worker/Job.pm]: done

File: lib/OpenQA/WebSockets/Controller/Worker.pm [hunk 3]
```diff
@@ -120,7 +121,7 @@ sub _message {
         my $wid                   = $worker->{id};
 
         $worker_status->{$wid} = $json;
-        log_debug(sprintf('Received from worker "%u" worker_status message "%s"', $wid, Dumper($json)));
+        log_debug(sprintf('Received from worker "%u" worker_status message "%s"', $wid, YAML::XS::Dump($json)));
```
Comment (kraih) [inline lib/OpenQA/WebSockets/Controller/Worker.pm]: Unrelated change?

File: lib/OpenQA/Worker/Job.pm [hunk 4]
```diff
@@ -660,35 +662,37 @@ sub _upload_results {
...
         worker_id             => $worker_id,
         cmd_srv_url           => $job_url,
         worker_hostname       => $global_settings->{WORKER_HOSTNAME},
-        test_execution_paused => $status_from_os_autoinst->{test_execution_paused} // 0,
+        test_execution_paused => 0,
     );
 
+    my $test_status = {};
+    if (-r $status_file) {
+        $test_status = Cpanel::JSON::XS->new->decode(path($status_file)->slurp);
```
Comment (kraih) [inline lib/OpenQA/Worker/Job.pm]: Just noticed that this class already imports `decode_json`, that shortens the code even more. 😉

Comment (kraih) [reply lib/OpenQA/Worker/Job.pm]: (just in case you didn't know, `Mojo::JSON` uses `Cpanel::JSON::XS` automatically when available)

File: lib/OpenQA/WebSockets/Controller/Worker.pm (hunk 3 above)
Comment (perlpunk) [reply lib/OpenQA/WebSockets/Controller/Worker.pm]: Ah yes, kind of.
I was debugging and comparing the logfiles with `vimdiff`. Data::Dumper output is by default unsorted and not well suited for a diff, IMHO.
I can take it out if you want.

File: lib/OpenQA/Worker/Job.pm (hunk 4 above)
Comment (perlpunk) [reply lib/OpenQA/Worker/Job.pm]: oh, and I realized that I never added a `use Cpanel::JSON::XS`

File: lib/OpenQA/WebSockets/Controller/Worker.pm (hunk 3 above)
Comment (kraih) [reply lib/OpenQA/WebSockets/Controller/Worker.pm]: True, that's one of the reasons [Mojo::Util::dumper](https://github.com/mojolicious/mojo/blob/93606e4d453bb29fddae2ad296017ad9b50d8531/lib/Mojo/Util.pm#L141) exists. I have no preference, others might.

Comment (kalikiana) [reply lib/OpenQA/WebSockets/Controller/Worker.pm]: I've done the same kind of temporary change before. I think it's excellent for debugging.

**Comment by okurz** [comment]
the description still mentions "Work in progress"?

**Comment by perlpunk** [comment]
Rebased and squashed, removed WIP from description

File: lib/OpenQA/WebSockets/Controller/Worker.pm (hunk 3 above)
Comment (perlpunk) [reply lib/OpenQA/WebSockets/Controller/Worker.pm]: I repeated the benchmark with a bigger hash and now times are almost equal:
```
perl -wE'
use YAML::XS ();
use Data::Dumper;
use Benchmark;
my $hash = YAML::XS::LoadFile("worker-status.yaml");
timethese($ARGV[0] || -1, {
  yaml => sub { YAML::XS::Dump($hash) },
  dumper => sub { Data::Dumper::Dumper($hash) },
});' 50000
Benchmark: timing 50000 iterations of dumper, yaml...
    dumper:  3 wallclock secs ( 2.44 usr +  0.00 sys =  2.44 CPU) @ 20491.80/s (n=50000)
      yaml:  2 wallclock secs ( 2.42 usr +  0.00 sys =  2.42 CPU) @ 20661.16/s (n=50000)
```
`worker-status.yaml`:
```yaml
---                                                                         
current_webui_host: http://localhost:47771
job:
  URL: http://localhost:20013/G9bUHr5eIMESFMlw
  assets:
    iso:
    - Core-7.2.iso
  assigned_worker_id: 1
  blocked_by_id: ~
  clone_id: ~
  group_id: ~
  id: 1
  name: tinycore-1-flavor-i386-Build1-core@coolone
  priority: 50
  result: none
  settings:
    ARCH: i386
    BUILD: '1'
    CDMODEL: ide-cd
    DISTRI: tinycore
    FLAVOR: flavor
    HDDMODEL: ide-drive
    INTEGRATION_TESTS: '1'
    ISO: Core-7.2.iso
    JOBTOKEN: G9bUHr5eIMESFMlw
    LOG_LEVEL: debug
    MACHINE: coolone
    NAME: 00000001-tinycore-1-flavor-i386-Build1-core@coolone
    OPENQA_HOSTNAME: localhost:47771
    PUBLISH_HDD_1: core-hdd.qcow2
    QEMU: i386
    QEMUPORT: 20012
    QEMU_NO_FDC_SET: '1'
    QEMU_NO_KVM: '1'
    QEMU_NO_TABLET: '1'
    RETRY_DELAY: 5
    RETRY_DELAY_IF_WEBUI_BUSY: 60
    TEST: core
    TESTING_ASSERT_SCREEN_TIMEOUT: '1'
    VERSION: '1'
    VNC: '91'
    WORKER_CLASS: qemu_i386,qemu_x86_64
    WORKER_HOSTNAME: 127.0.0.1
  state: scheduled
  t_finished: ~
  t_started: ~
  test: core
status: working
```
